pkg/fixedpoint/convert.go

```diff
@@ -164,47 +225,22 @@ func (v Value) MarshalJSON() ([]byte, error) {
 }
 
 func (v *Value) UnmarshalJSON(data []byte) error {
-	var a interface{}
-	var err = json.Unmarshal(data, &a)
-	if err != nil {
-		return err
+	if bytes.Compare(data, []byte{'n', 'u', 'l', 'l'}) == 0 {
+		*v = Zero
+		return nil
 	}
-
-	switch d := a.(type) {
-	case float64:
-		*v = NewFromFloat(d)
-
-	case float32:
-		*v = NewFromFloat32(d)
-
-	case int:
-		*v = NewFromInt(d)
-
-	case int64:
-		*v = NewFromInt64(d)
-
-	case string:
-		v2, err := NewFromString(d)
-		if err != nil {
-			return err
-		}
-
-		*v = v2
-
-	default:
-		return fmt.Errorf("unsupported type: %T %v", d, d)
-
+	if len(data) == 0 {
+		*v = Zero
+		return nil
 	}
-
-	return nil
-}
-
-func Must(v Value, err error) Value {
-	if err != nil {
-		panic(err)
+	var err error
+	if data[0] == '"' {
+		data = data[1 : len(data)-1]
```

Comment:
You can do this

```
var a interface{}
err := json.Unmarshal(data, &a)
...
switch a.(type) {
    case string:
    case int64:
}
```

Reply:
json.Unmarshal will accidentally convert numbers to float32, and lost the precision.  
directly parse from string would be better (well, maybe not in the legacy fixedpoint, just to make sure it doesn't fail)

Reply:
json.Unmarshal now cast number to float64, so it should be fine. if we really need high precision, we can write the number in the string form.